spec-cordova/prepare.spec.js

```diff
@@ -17,175 +17,265 @@
+                    done();
+                });
+            });
         });
-        it('Test 002 : should not run inside a cordova-based project with no platforms', function(done) {
-            list_platforms.and.returnValue([]);
-            prepare().then(function() {
-                expect('this call').toBe('fail');
-            }, function(err) {
-                expect('' + err).toContain('No platforms added to this project. Please use `cordova platform add <platform>`.');
-            }).fin(done);
+
+        describe('success', function () {
+            beforeEach(function () {
+                spyOn(util, 'cdProjectRoot').and.returnValue(project_dir);
+                spyOn(util, 'preProcessOptions').and.callFake(function(options) {
+                    let platforms = options.platforms || [];
+                    return {'platforms':platforms};
+                });
+            });
+            it('should fire the before_prepare hook and provide platform and path information as arguments', function(done) {
+                prepare({}).then(function() {
+                    expect(HooksRunner.prototype.fire.calls.argsFor(0)[0]).toEqual('before_prepare');
+                    done();
+                });
+            });
+            it('should invoke restore module\'s installPlatformsFromConfigXML method', function(done) {
+                prepare({}).then(function() {
+                    expect(restore.installPlatformsFromConfigXML).toHaveBeenCalled();
+                    done();
+                }).fail(function(err){
+                    expect(err).toBeUndefined();
+                    done();
+                });
+            });
+            it('should retrieve PlatformApi instances for each platform provided', function(done) {
+                prepare({'platforms':['android', 'ios']}).then(function() {
+                    expect(platforms.getPlatformApi.calls.count()).toBe(4);
+                    expect(platforms.getPlatformApi.calls.argsFor(0)[0]).toBe('android');
+                    expect(platforms.getPlatformApi.calls.argsFor(0)[1]).toBe('/some/path/platforms/android');
+                    expect(platforms.getPlatformApi.calls.argsFor(1)[0]).toBe('ios');
+                    expect(platforms.getPlatformApi.calls.argsFor(1)[1]).toBe('/some/path/platforms/ios');
+                    expect(platforms.getPlatformApi.calls.argsFor(2)[0]).toBe('android');
+                    expect(platforms.getPlatformApi.calls.argsFor(3)[0]).toBe('ios');
+                    done();
+                }).fail(function(err){
+                    expect(err).toBeUndefined();
+                    done();
+                });
+            });
+            it('should invoke restore module\'s installPluginsFromConfigXML method', function(done) {
+                prepare({platforms:[]}).then(function() {
+                    expect(restore.installPluginsFromConfigXML).toHaveBeenCalled();
+                    done();
```

Comment:
Sounds good. Should I add a `.fail( function (e) { ...` to all of the tests? I added it selectively throughout the tests. 

Also, didn't know about `fail('unexpected failure handler invoked');`. Is that a built in jasmine function? Doesn't seem documented on their site.